### ISSUE TITLE
`COMException` 0x80004004 during `EnsureCoreWebView2Async`

### ISSUE THREAD

**Issue by exchange321** [MicrosoftEdge/WebView2Feedback#2410]
**Description**
We have noticed that some users encountered the following error when initializing WebView2
```cs
System.Runtime.InteropServices.COMException: Operation aborted (Exception from HRESULT: 0x80004004 (E_ABORT))
  ?, in void Marshal.ThrowExceptionForHRInternal(int errorCode, IntPtr errorInfo)
  ?, in async Task<CoreWebView2Controller> CoreWebView2Environment.CreateCoreWebView2ControllerAsync(IntPtr ParentWindow)
  ?, in async Task WebView2.EnsureCoreWebView2Async(CoreWebView2Environment environment)+Init(?)
  ?, in async void MainViewModel.Initialize(WebView2 webView2)
  ?, in void AsyncMethodBuilderCore.ThrowAsync(Exception exception, SynchronizationContext targetContext)+(object state) => { }
  ?, in object ExceptionWrapper.InternalRealCall(Delegate callback, object args, int numArgs)
  ?, in object ExceptionWrapper.TryCatchWhen(object source, Delegate callback, object args, int numArgs, Delegate catchHandler)
```
`MainViewModel.Initialize` is called when `UserControl.Loaded` is triggered.

Not sure if it helps, we also observed that for some reasons the WebView2 element is already disposed when being initialized:
```cs
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'WebApp'.
  ?, in void WebView2.VerifyNotDisposed()
  ?, in CoreWebView2 WebView2.get_CoreWebView2()
  ?, in async void MainViewModel.LoadWebApp()
  ?, in void AsyncMethodBuilderCore.ThrowAsync(Exception exception, SynchronizationContext targetContext)+(object state) => { }
  ?, in object ExceptionWrapper.InternalRealCall(Delegate callback, object args, int numArgs)
  ?, in object ExceptionWrapper.TryCatchWhen(object source, Delegate callback, object args, int numArgs, Delegate catchHandler)
```
`MainViewModel.LoadWebApp` is called when `WebView2.CoreWebView2InitializationCompleted` is triggered.

**Version**
SDK: 1.0.1020.30
Runtime: >=95.0.1020.30
Framework: WPF (.NET Framework 4.8)
OS: >=10.0.22000.0

**Repro Steps**
We're unable to reproduce the issue but we can observe the occurrences of ^ from our logs.

**Additional context**
I'm quite new to Windows app development so I might have done something wrong here. We have a `Dispose` function:
```cs
        protected override void Dispose(bool cleanManagedResources)
        {
            if (!cleanManagedResources)
            {
                return;
            }

            if (WebView2 != null)
            {
                WebView2.Dispose();
            }
        }
```
which, when called, disposes the WebView2 element. It will be triggered when the user closes the application. Could it be that there is some race condition between the WebView2 being disposed and the initialization still running/the event still being dispatched?

[AB#39536250](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/39536250)

**Comment by LiangTheDev** [comment]
E_ABORT means that the creation of WebView is cancelled. It could happen if the app window got closed by the user while we are creating the WebView.
You could catch exceptions from EnsureCoreWebView2Async to avoid crashing, and log events to help figure whether the app is really closed by the user.

**Comment by exchange321** [comment]
@LiangTheDev thanks for looking into this. I'll add some logging around ^ to see if that's the issue 👍

**Comment by fdzatdtv** [comment]
We also observe COMException 0x80004004. We tried to catch the exception and decorate it. But it seems that the exception is thrown in separate thread and our catch clause isn't run. We don't see the decorated exception from our logs.
Here our try/catch code:
```
private async void UserControl_Load(object sender, EventArgs e)
{
    Uri uri = null;
    try
    {
        uri = await InitializeAsync().ConfigureAwait(true);
    }
    catch (Exception ex)
    {
        throw DecorateException(ex);
    }
}

private async Task<Uri> InitializeAsync()
{
    // some work
    try
    {
        uri = await InitializeAsync(this.StartUri).ConfigureAwait(true);
    }
    finally
    {
        // some work
    }
}

private async Task<Uri> InitializeAsync(Uri startUri)
{
    var environment = await CoreWebView2Environment.CreateAsync(...).ConfigureAwait(true);

    Task localInitTask = webView.EnsureCoreWebView2Async(environment);
    this._initTask = localInitTask;

    if (localInitTask != null)
    {
        await localInitTask.ConfigureAwait(true);
    }

    return startUri;
}
```

**Stacktrace**
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
   at Microsoft.Web.WebView2.Core.CoreWebView2Environment.<CreateCoreWebView2ControllerAsync>d__54.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Microsoft.Web.WebView2.WinForms.WebView2.<InitCoreWebView2Async>d__18.MoveNext()

and
    at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
    at Microsoft.Web.WebView2.Core.CoreWebView2Environment.<CreateCoreWebView2ControllerAsync>d__54.MoveNext()
    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(Object stateMachine)
    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
    at Microsoft.Web.WebView2.Core.CoreWebView2CreateCoreWebView2ControllerCompletedHandler.Invoke(Int32 errCode, ICoreWebView2Controller createdController)
    at System.Windows.Forms.UnsafeNativeMethods.IntDestroyWindow(HandleRef hWnd)
    at System.Windows.Forms.UnsafeNativeMethods.DestroyWindow(HandleRef hWnd)
    at System.Windows.Forms.NativeWindow.DestroyHandle()
    at System.Windows.Forms.Control.DestroyHandle()
    at System.Windows.Forms.Control.Dispose(Boolean disposing)


**Environment**
Evergreen Runtime 101.0.1210.39 (SDK: 1.0.1054.31)
Application: WinForms (x86)
OS: Windows Server 2016 Standard

**Comment**
Catching the COMException at the global exception handler is not an option. We would like to catch it as close as possible.

**Comment by LiangTheDev** [comment]
We should ensure that the exception is propagated back to the original calling context. Tracked the issue internally.

**Comment by ElyssaJyu** [comment]
We are tracking the issue but cannot reproduce it. Can you provide the sample code?

**Comment by ElyssaJyu** [comment]
I will close this issue as no response for a long while. If there is still an issue, please reactivate it.

**Comment by Aadithan** [comment]
@exchange321 I am getting a similar exception. How did you resolve it? Can you please post it here?

**Comment by ElyssaJyu** [comment]
Hey @Aadithan, could you describe your workaround and the exception?

**Comment by ElyssaJyu** [comment]
Hey @Aadithan, E_ABORT will happen if the parent window is being destroyed before creation was finished. This may be caused by closing the app before webview2 control initialization completed. 
A potential workaround: create and initialize a webview2, then call its Dispose () immediately. 
Since webview2 initialization is async. This may be able to repro the exception. Could you try to test it in your app to check whether the exception will happen?

**Comment by breakeddev** [comment]
I have added webview2 in a windows credential provider in c++,  Webview2 load successfully if system is restarted or logged out. One thing failing is if end-user locked its screen and try to load webview2 it fails. No error is catched in try catch block. On complete event the same error code 0x80004004.
tried to move user directory to temporary locations with all rights, tried to wait for 10 seconds before navigation but all is failed. Tried to provide browser executable folder as well.

**Comment by ElyssaJyu** [comment]
Hey @breakeddev, do you mean you try to load webview2 on user lock screen? Webview2 won't be allowed to be used for credential providers, so WebView2 doesn't allow running as System account. This is a Windows desired limitation due to potential security concerns.

**Comment by ElyssaJyu** [comment]
Hey @breakeddev, 

> It did worked on system restart and if user is logged out completely. If end user is locked it doesnot load

It seems this was working until the browser put in a check for running as SYSTEM and then erroring on purpose. To my knowledge, the flag "--allow-run-as-system" seems to be retained, but currently building credential providers with webview2 is not supported due to security concerns. 
This [GH thread](https://github.com/MicrosoftEdge/WebView2Feedback/issues/1907#issuecomment-1372525335) includes more detailed information about it. Also, is your scenario similar to [this issue](https://github.com/MicrosoftEdge/WebView2Feedback/issues/2868), like building a custom UI?

**Comment by breakeddev** [comment]
@ElyssaJyu  thanks for the threads information, the error messages are not similar and i did already gone through those links. 
I have verified that in both modes the library loads as "SYSTEM" user.  Regarding your opinion that "browser put in a check after lock", i have the below login scenarios 
Login -> Lock Screen -> WebView2 don't load
Login -> Lock Screen -> Failure  (WebView2 don't load) -> Relogin -> Logout -> Webview2 loads successfully 
Login -> Logout -> webview2 loads successfully 
Without login -> Webview2 loads successfully 

My scenario is to load web browser in a window on login screen and open specific links, In webview context menus and every thing should be disabled.

**Comment by LiangTheDev** [comment]
As indicated in the thread, E_ABORT most likely to happen if the parent hwnd of WebView2 is closed while we are initializing WebView2. If you have a consistent repro, could you please take an ETW trace (https://github.com/MicrosoftEdge/WebView2Feedback/blob/main/diagnostics/etw.md) for us to do an initial triage of the issue? Or, provide a simple repro project with repro steps.

**Comment by LiangTheDev** [comment]
@G-Equipment, you could email me at [lzhao@microsoft.com](mailto:lzhao@microsoft.com).

Could you also try to wrap EnsureCoreWebView2Async in try - catch and see whether that makes any difference (not crashing the app)?

**Comment by LiangTheDev** [comment]
The ETW events indicate that the internal WebView2 hwnd(which is a child hwnd of the winforms control) did receive WM_NCDESTROY right after it is created. It is unlikely that the Winforms page/control is destroyed. So, something else is destroying the WebView2 hwnd.

I noticed that the app is also using CEF, which also uses Chromium code. I am wondering whether it is possible that the CEF part is mistaken WebView2's hwnd as its own and destroyed it. Is it possible to run the same test without CEF part of the app running?

**Comment by LiangTheDev** [comment]
If it didn't work before using CEF, then it should not be related to CEF. I am out of ideas.

Since the hwnd is destroyed, we would have to poke around that, like only call Load when visibility is true when VisibilityChanged event is received, add event handler for HandleCreated and HandleDestroyed events to see whether these events fired during WebView initialization.

**Comment by LiangTheDev** [comment]
Something like this?
```
            this.VisibleChanged += WebBrowser_VisibilityChanged;
            this.HandleDestroyed += WebBrowser_HandleDestroyed;
            this.HandleCreated += WebBrowser_HandleCreated;

        private void WebBrowser_HandleCreated(object sender, EventArgs e)
        {
             Trace.WriteLine("Debug HandleCreated");
        }
        private void WebBrowser_HandleDestroyed(object sender, EventArgs e)
        {
             Trace.WriteLine("Debug HandleDestroyed");
        }
        private void WebBrowser_VisibilityChanged(object sender, EventArgs e)
        {
             Trace.WriteLine("Debug VisibilityChanged, Visible: " + Visible);
             if (Visible)
             {
                 Load();
             }
        }
 
```

**Comment by lightheart33** [comment]
**@breakeddev**
With version 122.0.2365.92, I am having the exact same scenario as you described, and the exact same behavior with a twist:
When trying on a locked screen to create the WebView2 environment, it apparently never completes.
and inside the logged on user, a dialog appears:
![image](https://github.com/MicrosoftEdge/WebView2Feedback/assets/107999791/3c8a5976-424d-4639-b14b-378caeab83d8)
First when I press OK inside the user session, does the Prelogon WebView2 creation continue with an error.
Its a bit strange, because as far as I know the Microsoft Edge has its own WebView2 installation,
and I see no reason why this should interfere, but something does.
Ok the logged on user does not have permission to write under C:\Windows\TEMP, 
therefore I have under the CSIDL_COMMON_APPDATA folder created a subfolder, where all users can write,
and it does seem to help, at the moment I am not getting the Microsoft Edge error dialog inside the user session.

However, occasionally I am still getting the 0x80004004 (E_ABORT) on the locked screen scenario.
Have you discovered anything new about this issue? possible a workaround? :)

**Comment by lightheart33** [comment]
@breakeddev 
WebView2 is already embedded in an application, although implemented in another language, but that does not seem to help anything. Perhaps there is something with the visibility I have missed, and which is causing an issue, but somehow I don't think so.

Since I wrote I also attempted finding out the already logged on user, and obtaining an access token for this user.
Then as System user (Windows Prelogon) I first performed an impersonate of the logged on user. and executed the application 
with WebView2 as this user, but I still get 0x80004004. As far as I have been able to google, when two processes try to 
use the same EBWebView folder, then the second timeout after a couple of minutes, with 0x80004004, which I was able to reproduce. However the logged on user hadn't started any WebView2 or MsEdge process, although I first I had to terminate all the MsEdge prelaunch processes. I suspect the EBWebView folder contain somewhere a file, that is either locked by another process, 
or contain data about the lock, preventing a second process from using WebView2 (EBWebView folder) at the same time.
